launcher/container_runner.go

```diff
@@ -0,0 +1,318 @@
+				"format":   {"full"},
+			}.Encode(),
+		}
+		idToken, err := mdsClient.Get(u.String())
+		if err != nil {
+			return nil, fmt.Errorf("failed to get principal tokens: %w", err)
+		}
+		return [][]byte{[]byte(idToken)}, nil
+	}
+
+	return &ContainerRunner{
+		container,
+		launchSpec,
+		conn,
+		CreateAttestationAgent(tpm, client.AttestationKeyECC, conn, principalFetcher),
+	}, nil
+}
+
+// parseEnvVars parses the environment variables to the oci format
+func parseEnvVars(envVars []spec.EnvVar) []string {
+	var result []string
+	for _, envVar := range envVars {
+		result = append(result, envVar.Name+"="+envVar.Value)
+	}
+	return result
+}
+
+// appendTokenMounts appends the default mount specs for the OIDC token
+func appendTokenMounts(mounts []specs.Mount) []specs.Mount {
+	m := specs.Mount{}
+	m.Destination = containerTokenMountPath
+	m.Type = "bind"
+	m.Source = HostTokenPath
+	m.Options = []string{"rbind", "ro"}
+
+	return append(mounts, m)
+}
+
+// measureContainerClaims will measure various container claims into the COS
+// eventlog in the AttestationAgent.
+func (r *ContainerRunner) measureContainerClaims(ctx context.Context) error {
+	image, err := r.container.Image(ctx)
+	if err != nil {
+		return err
+	}
+	if err := r.attestAgent.MeasureEvent(cel.CosTlv{cel.ImageRefType, []byte(image.Name())}); err != nil {
+		return err
+	}
+	if err := r.attestAgent.MeasureEvent(cel.CosTlv{cel.ImageDigestType, []byte(image.Target().Digest)}); err != nil {
+		return err
+	}
+	r.attestAgent.MeasureEvent(cel.CosTlv{cel.RestartPolicyType, []byte(r.launchSpec.RestartPolicy)})
+	if imageConfig, err := image.Config(ctx); err == nil { // if NO error
+		if err := r.attestAgent.MeasureEvent(cel.CosTlv{cel.ImageIDType, []byte(imageConfig.Digest)}); err != nil {
+			return err
+		}
+	}
+
+	contianerSpec, err := r.container.Spec(ctx)
+	if err != nil {
+		return err
+	}
+	for _, arg := range contianerSpec.Process.Args {
+		if err := r.attestAgent.MeasureEvent(cel.CosTlv{cel.ArgType, []byte(arg)}); err != nil {
+			return err
+		}
+	}
+	for _, env := range contianerSpec.Process.Env {
+		if err := r.attestAgent.MeasureEvent(cel.CosTlv{cel.EnvVarType, []byte(env)}); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// Run the container
+// Doesn't support container restart yet
+// Container output will always be redirected to stdio for now
+func (r *ContainerRunner) Run(ctx context.Context) error {
+	if err := r.measureContainerClaims(ctx); err != nil {
+		return fmt.Errorf("failed to measure container claims: %v", err)
+	}
+
+	// Create hostTokenPath directory. Should come before NewTask.
+	token, err := r.attestAgent.Attest(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to retrieve attestation service token: %v", err)
+	}
+
+	// Write OIDC token for the container.
+	if err := os.MkdirAll(HostTokenPath, 0644); err != nil {
+		log.Fatal(err)
+	}
+	err = os.WriteFile(path.Join(HostTokenPath, attestationVerifierTokenFile), token, 0644)
```

Comment:
Can combine two lines `if err = os.WriteFile...; err != nil {`

Reply:
I think we should fix this in a subsequent pr.